validator/client/src/main/java/tech/pegasys/teku/validator/client/proposerconfig/AbstractProposerConfigProvider.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright 2022 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.validator.client.proposerconfig;
+
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import tech.pegasys.teku.infrastructure.async.AsyncRunner;
+import tech.pegasys.teku.infrastructure.async.SafeFuture;
+import tech.pegasys.teku.validator.client.ProposerConfig;
+import tech.pegasys.teku.validator.client.proposerconfig.loader.ProposerConfigLoader;
+
+public abstract class AbstractProposerConfigProvider implements ProposerConfigProvider {
+  private static final Logger LOG = LogManager.getLogger();
+
+  private final boolean refresh;
+  private final AsyncRunner asyncRunner;
+  protected final ProposerConfigLoader proposerConfigLoader;
+  private Optional<ProposerConfig> lastProposerConfig = Optional.empty();
```

Comment:
My suggestion would be for this to be stored as a `SafeFuture` then if we already have a request in progress we can return this future to "merge" the two requests rather than having to throw an error if we don't already have a config. While we probably won't get concurrent requests much, I don't think there's anything that explicitly prevents them happening so we probably want to just smoothly handle it.

It would also need to be `volatile` since it's being accessed from multiple threads.

Reply:
I ended up synchronizing. I think it simplifies the handling a bit, without a huge cost (at the end the call frequency is very low).